keps/sig-node/20181106-in-place-update-of-pod-resources.md

```diff
@@ -211,17 +235,41 @@ new desired resources (i.e Spec.Containers[i].Resources.Requests) to the sum.
   UpdateContainerResources CRI API to update Container resource limits. Once
   all Containers are successfully updated, it updates
   Pod.Status.ContainerStatuses[i].Resources to reflect new resource values.
-* If new desired resources don't fit, Kubelet rejects the resize, and no
-  further action is taken.
-  - Kubelet retries the Pod resize at a later time.
+* If new desired resources don't fit, Kubelet will not act on the resize, and
+  no further action is taken at this time. Kubelet will retry the resize at a
+  later time.
+
+<<[UNRESOLVED]>>
+As discussed in comments, "no signal" is a poor signal.  I advocate for at
+least this:
 
-If multiple Pods need resizing, they are handled sequentially in the order in
-which Pod additions and updates arrive at Kubelet.
+"""
+Kubelet will generate Events on the Pod whenever a resize is acted upon or when
+a resize is deferred.  This will allow humans to know that progress is being
+made.
+"""
+
+I also advocate for this, though it can be deferred (though I bet we need it
+eventually):
+
+"""
+When a resize is in progress, Kubelet will create a Condition on the pod,
```

Comment:
Staring a new thread on (hopefully) the last contentious topic.

After the agreement above, the only way that a VPA-like entity can ascertain whether a resource update is being acted on is to look at `status.resourcesAllocated` - if it has been updated to match `spec.resources` then the resize is "in progress".  If it has not been updated either:
1) Kubelet can't actuate it right now
2) Kubelet can't actuate it ever
3) Kubelet is out-to-lunch

I propose that we do better.  I suggested conditions, but you shot that full of holes :P

How about a map of resource name to resize-status.  E.g. `status.resize.  Any time a write happens to `spec.resources, including create (?), this map is populated with one entry for each changed element of `spec.resources` with value "Unknown".  Kubelet, on its next status update, will update the map values to either "InProgress", "Deferred", or "Infeasible".  When a resize is complete, kubelet must remove the map key.

This gives a more precise signal to resize-actors.

There's still a race where a resize actor could send a resize, kubelet starts the update to InProgress, but the resize actor sends another resize before kubelet completes the update. An observer could see the second `spec` but the status from the first update.  Kubelet would need to use RV to make sure that doesn't happen.

I'm very open to other (simpler) ideas, but timing out seems like a bad choice to me

Reply:
IIUC, you have two asks:
1. I want an ACK that says that the kubelet saw the resize so I know the kubelet isn't out to lunch.  Hence the "Unknown" initial value.
2. I want to be able to differentiate between "deferred" and "infeasible".

For (1), I think a single observedGeneration field would be enough, and would provide a mechanism for you to know if the kubelet has looked at and tried to reconcile any version of an object.  Although I don't find the need for this _that_ compelling.  We haven't needed an ack for other changes to pods such as image updates.

Unless I'm missing something, "deferred" and "infeasible" are the same when waiting for resourcesAllocated to be updated.  If I interpret "infeasible" as "Kubelet can't actuate it ever, unless something else leaves the node", then that is the same as 1.  If I interpret 2 as "Kubelet can't actuate it ever, no matter what", then that isn't possible in the current design, as it will continue to retry the feasibility check.

Taking a step back, the resourcesAllocated field lets you distinguish between "its waiting on feasibility" and "its waiting on actuation".  I _think_ this is what we want, right?  Or is there something else missing?

Reply:
Forgot to mention: I'd prefer a mechanism that doesn't have multiple writers to a single field.

I agree it is possible to use resource version to prevent the race, but it means either the kubelet just skips conflicting updates entirely, or has to recompute the status based on the new spec.Resources, which would require significant refactoring.  The flow today is roughly:

syncPod triggered by pod update, runtime state update, or periodic -> get v1.Pod from watch cache &&  get container status from CRI ->  generate new status -> store new status so it can be updated in the background -> call CRI to make changes.

We probably want to just skip those updates, given the new generation of the pod should already be coming through the pipe.  But IIUC, it means any other pod status updates we want to do don't get done either.

Reply:
I agree with @dashpole . ObservedGeneration can be used to pin-point the spec that the condition applies to, and the Reason + Message fields can describe the precise cause for the condition. This looks stateless and reconstructible, and is simple.

Reply:
@dashpole: 

> IIUC, you have two asks:
> 
>   1. I want an ACK that says that the kubelet saw the resize so I know the kubelet isn't out to lunch. Hence the "Unknown" initial value.
>   2. I want to be able to differentiate between "deferred" and "infeasible".
>
> For (1), I think a single observedGeneration field would be enough, and would provide a mechanism for you to know if the kubelet has looked at and tried to reconcile a generation of an object. TBH, I don't find the need for this that compelling. We haven't needed an ack for other changes to pods such as image updates. But i'm alright with it if we think it should be a requirement.

Not exactly.  The point of Unknown (and specifically the automatic setting of
it) is to avoid internally-inconsistent information.  Consider:

- pod has `cpu: 1`
- user changes `cpu: 2`
- kubelet observes that and writes "InProgress"
- user changes `cpu: 8`
- observer sees that a resize to 8 is in progress

By making it automatically change to Unknown, the observer has to wait for
kubelet to accept, defer, or reject.  No ambiguity.

I'm not sure observedGeneration can cover that?  Am I missing something?

> If I interpret "infeasible" as "Kubelet can't actuate it ever, unless something else leaves the node",

I called that deferred.  It's theoretically feasible but "not right now".

> If I interpret 2 as "Kubelet can't actuate it ever, no matter what", then that isn't possible in the current design, as it will continue to retry the feasibility check.

This is what I meant by infeasible.  For example "this node doesn't have enough
CPU to *ever* satisfy that request".  Had it gone thru the scheduler it would
not have been scheduled to this node.  It seems low-probability that the
scheduler would screw that up, but now that resource updates bypass the
scheduler, it seems much more likely for a user or auto-scaler to trigger it.
Kubelet should be smart enough to know "this update CAN'T WORK".

> Taking a step back, the resourcesAllocated field lets you distinguish between "its waiting on feasibility" and "its waiting on actuation".

If resourcesAllocated has been updated, I know that kubelet is actuating "right
now".  If it has not been updated, I have no idea if it is waiting for a pod to
finish (might be worth waiting a while) or if it will NEVER happen (in which
case waiting for a timeout is utterly pointless).  I actually considered an
"ETA" field but that seemed like diminishing returns

Autoscaling is going to be a critical piece of the scaling strategy for apps.
Anything we can do to make it faster and more deterministic is worthwhile.
This is why I am so uptight here.

@vinaykul

> ObservedGeneration can be used to pin-point the spec that the condition applies to, and the Reason + Message fields can describe the precise cause for the condition. This looks stateless and reconstructible, and is simple.

Maybe I am missing something, can you spell it out for me a bit more?

Reply:
>>@vinaykul

>>ObservedGeneration can be used to pin-point the spec that the condition applies to, and the Reason + Message fields can describe the precise cause for the condition. This looks stateless and reconstructible, and is simple.

>Maybe I am missing something, can you spell it out for me a bit more?

One possible way:
- We can set PodCondition.Status to True or False based on whether the last observed resize request is admissible or not.
- If false:
  - We set the Reason field to state a reason e.g. InsufficientNodeCapacity.
  - We set the Message field to state something like "Pod ResourceVersion=X, Generation=Y resize not accepted, not enough CPU"
- If true, Reason=Successful, (and optional)Message="CPU resize was accepted"

Does that help clarify?